### PR TITLE
terraform-provider: add input validation

### PR DESCRIPTION
**Pull request by msanft**
### Context

There are a lot of inputs that we can validate without further configuration in the Terraform provider, to catch errors early.

### Proposed change(s)

- Add input validation for `constellation_image` and `constellation_cluster`. `constellation_attestation` already has input validation.

### Additional info

- [AB#3613](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3613)
- [E2E Test](https://github.com/edgelesssys/constellation/actions/runs/7273881323)

### Checklist

- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: terraform-provider-constellation/internal/provider/cluster_resource.go [hunk 1]
```diff
@@ -331,6 +360,69 @@ func (r *ClusterResource) Schema(_ context.Context, _ resource.SchemaRequest, re
...
+func (r *ClusterResource) ValidateConfig(ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse) {
+	var data ClusterResourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Azure Config is required for Azure
+	tflog.Info(ctx, fmt.Sprintf("CSP: %s", data.CSP.ValueString()))
+	tflog.Info(ctx, fmt.Sprintf("Azure: %v", data.Azure.IsNull()))
```
Comment (daniel-weisse) [inline terraform-provider-constellation/internal/provider/cluster_resource.go:373]: nit: useless log statement

File: terraform-provider-constellation/internal/provider/shared_attributes.go [hunk 2]
```diff
@@ -166,6 +170,9 @@ func newImageAttributeSchema(t attributeType) schema.Attribute {
 				MarkdownDescription: "Semantic version of the image.",
 				Computed:            !isInput,
 				Required:            isInput,
+				Validators: []validator.String{
+					stringvalidator.RegexMatches(semverRegex, "must be a valid semantic version"),
```
Comment (daniel-weisse) [inline terraform-provider-constellation/internal/provider/shared_attributes.go:174]:
```suggestion
					stringvalidator.RegexMatches(semverRegex, "must be a valid semantic version with a v prefix in the form of vX.Y.Z"),
```
nit: "real" semantic versioning doesnt have the v prefix

**Review by daniel-weisse — APPROVED** [review]

File: terraform-provider-constellation/internal/provider/shared_attributes.go [hunk 3]
```diff
@@ -166,6 +170,9 @@ func newImageAttributeSchema(t attributeType) schema.Attribute {
 				MarkdownDescription: "Semantic version of the image.",
 				Computed:            !isInput,
 				Required:            isInput,
+				Validators: []validator.String{
```
Comment (elchead) [inline terraform-provider-constellation/internal/provider/shared_attributes.go:173]: we should also allow complete image references (e.g. ` "ref/main/stream/nightly/v2.14.0-pre.0.20231214193540-2c50abcc919b"`), which are needed for tests too

File: terraform-provider-constellation/internal/provider/image_data_source.go [hunk 4]
```diff
@@ -103,19 +107,48 @@ func (d *ImageDataSource) Schema(_ context.Context, _ datasource.SchemaRequest,
...
+	if !data.CSP.Equal(types.StringValue("azure")) && !data.MarketplaceImage.IsNull() {
+		resp.Diagnostics.AddAttributeWarning(
+			path.Root("marketplace_image"),
+			"Marketplace images are currently only supported on Azure", "When another CSP than Azure is used, setting 'marketplace_image' has no effect.",
+		)
+	}
+
+	// Version should be a valid semver or short path, if set
+	if !data.Version.IsNull() {
+		_, semverErr := semver.New(data.Version.ValueString())
```
Comment (elchead) [inline terraform-provider-constellation/internal/provider/image_data_source.go:141]: why do we have a regex check in the attribute and a semver check here?

File: terraform-provider-constellation/internal/provider/shared_attributes.go (hunk 3 above)
Comment (msanft) [reply terraform-provider-constellation/internal/provider/shared_attributes.go:173]: This semver check is superfluous with the other check, which checks if the reference is either a semver, or a versionsapi short path, like the one you've sent. Thanks!

File: terraform-provider-constellation/internal/provider/shared_attributes.go [hunk 5]
```diff
@@ -20,6 +22,8 @@ const (
 	attributeOutput attributeType = false
 )
 
+var semverRegex = regexp.MustCompile(`^v\d+\.\d+\.\d+(-pre.*)?$`)
```
Comment (elchead) [inline terraform-provider-constellation/internal/provider/shared_attributes.go:25]:
```suggestion
```

**Review by elchead — APPROVED** [review]
lgtm besides the 1 remaining nit

**Comment by elchead** [comment]
@msanft should we mention this to the user as feature?

**Comment by msanft** [comment]
> @msanft should we mention this to the user as feature?

IMO, this should rather be something transparent to the user. As a user of a Terraform provider, I would expect it to have some kind of input validation. So I'd avoid praising it as a feature

**Pull request merged**